src/common/bitfield_util.h

```diff
@@ -0,0 +1,306 @@
+
+  bool IsUnsigned() const noexcept { return type_ == Type::kUnsigned; }
+
+  uint8_t Bits() const noexcept { return bits_; }
+
+  Status SetBitsCount(uint8_t new_bits) {
+    if (!IsSupportedBitLengths(type_, new_bits)) {
+      return Status::NotOK;
+    }
+    bits_ = new_bits;
+    return Status::OK();
+  }
+
+  Status SetType(Type new_type) {
+    if (!IsSupportedBitLengths(new_type, bits_)) {
+      return Status::NotOK;
+    }
+    type_ = new_type;
+    return Status::OK();
+  }
+
+  std::string ToString() const noexcept { return (IsSigned() ? "i" : "u") + std::to_string(static_cast<int>(bits_)); }
+
+ private:
+  BitfieldEncoding() = default;
+
+  Type type_;
+  uint8_t bits_;
+};
+
+struct BitfieldOperation {
+  // see https://redis.io/commands/bitfield/ to get more details.
+  enum class Type : uint8_t { kGet, kSet, kIncrBy };
+
+  Type type;
+  BitfieldOverflowBehavior overflow{BitfieldOverflowBehavior::kWrap};
+  BitfieldEncoding encoding{BitfieldEncoding::Create(BitfieldEncoding::Type::kSigned, 32).GetValue()};
+  uint32_t offset;
+  // INCRBY amount or SET value
+  int64_t value;
+};
+
+namespace detail {
+// return true if overflow. Status is not ok iff calling BitfieldEncoding::IsSupportedBitLengths()
+// for given op return false.
+StatusOr<bool> SignedBitfieldPlus(uint64_t value, int64_t incr, uint8_t bits, BitfieldOverflowBehavior overflow,
+                                  uint64_t *dst);
+
+// return true if overflow. Status is not ok iff calling BitfieldEncoding::IsSupportedBitLengths()
+// for given op return false.
+StatusOr<bool> UnsignedBitfieldPlus(uint64_t value, int64_t incr, uint8_t bits, BitfieldOverflowBehavior overflow,
+                                    uint64_t *dst);
+}  // namespace detail
+
+// safe cast from unsigned to signed, without any bit changes.
+// see also "Integral conversions" on https://en.cppreference.com/w/cpp/language/implicit_conversion
+// If the destination type is signed, the result when overflow is implementation-defined until C++20
+inline int64_t CastToSignedWithoutBitChanges(uint64_t x) {
+  int64_t res = 0;
+  memcpy(&res, &x, sizeof(res));
+  return res;
+}
+
+class BitfieldValue {
+ public:
+  BitfieldValue(BitfieldEncoding encoding, uint64_t value) noexcept : encoding_(encoding), value_(value) {}
+
+  template <class T, std::enable_if_t<std::is_integral_v<T>, int> = 0>
+  bool operator==(T rhs) const {
+    return value_ == static_cast<uint64_t>(rhs);
+  }
+
+  template <class T>
+  friend bool operator==(T lhs, const BitfieldValue &rhs) {
+    return rhs == lhs;
+  }
+
+  BitfieldEncoding Encoding() const noexcept { return encoding_; }
+
+  uint64_t Value() const noexcept { return value_; }
+
+ private:
+  BitfieldEncoding encoding_;
+  uint64_t value_;
+};
+
+// return true if overflow. Status is not ok iff calling BitfieldEncoding::IsSupportedBitLengths()
+// for given op return false.
+inline StatusOr<bool> BitfieldPlus(uint64_t value, int64_t incr, BitfieldEncoding enc,
```

Comment:
This function needs more detailed document. It means like check "(value, enc) + incr"